/* Copyright 2020 The FedLearner Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";
import "fedlearner/common/common.proto";

package fedlearner.common;

message FileInfoList {
  repeated string files = 1;
  repeated int64 idx = 2;
}

message BatchInfo {
  bool finished = 1;
  int64 file_idx = 2;
  int64 batch_idx = 4;
}

message AllocateTaskRequest {
  int64 rank_id = 1;
}

message AllocateTaskResponse {
  Status status = 1;
  FileInfoList file_info = 2;
}

message FinishFilesRequest {
  repeated int64 file_idx = 1;
}

message FinishFilesResponse {
  Status status = 1;
  string err_msg = 2;
}

service TransmitterMasterService {
  rpc AllocateTask(AllocateTaskRequest) returns (AllocateTaskResponse) {}
  rpc FinishFiles(FinishFilesRequest) returns (FinishFilesResponse) {}
}

message SyncRequest {
  int64 file_idx = 1;
}

message SyncResponse {}

message RecvFileFinishRequest {
  int64 file_idx = 1;
}

message RecvFileFinishResponse {}

message RecvTaskFinishRequest {}

message RecvTaskFinishResponse {}

message DataFinishRequest {}

message DataFinishResponse {}

message Request {
  BatchInfo batch_info = 1;
  bytes payload = 5;
}

message Response {
  Status status = 1;
  BatchInfo batch_info = 2;
  bytes payload = 4;
}

service TransmitterWorkerService {
  rpc RecvFileFinish(RecvFileFinishRequest) returns (RecvFileFinishResponse) {}
  rpc RecvTaskFinish(RecvTaskFinishRequest) returns (RecvTaskFinishResponse) {}
  rpc Transmit(stream Request) returns (stream Response) {}
  rpc DataFinish(DataFinishRequest) returns (DataFinishResponse) {}
}
